examples/sonoff_basic/main.c

```diff
@@ -0,0 +1,138 @@
+ * a simple $5 Sonoff Basic using HomeKit.
+ * The esp-wifi-config library is also used in this
+ * example. This means you don't have to specify
+ * your network's SSID and password before building.
+ *
+ * In order to flash the sonoff basic you will have to
+ * have a 3,3v (logic level) FTDI adapter.
+ *
+ * To flash this example connect 3,3v, TX, RX, GND
+ * in this order, beginning in the (square) pin header
+ * next to the button.
+ * Next hold down the button and connect the FTDI adapter
+ * to your computer. The sonoff is now in flash mode and
+ * you can flash the custom firmware.
+ *
+ * WARNING: Do not connect the sonoff to AC while it's
+ * connected to the FTDI adapter! This may fry your
+ * computer and sonoff.
+ *
+ */
+
+#include <stdio.h>
+#include <esp/uart.h>
+#include <esp8266.h>
+#include <FreeRTOS.h>
+#include <task.h>
+
+#include <homekit/homekit.h>
+#include <homekit/characteristics.h>
+#include <wifi_config.h>
+
+// The GPIO pin that is connected to the relay on the Sonoff Basic.
+const int relay_gpio = 12;
+// The GPIO pin that is connected to the LED on the Sonoff Basic.
+const int led_gpio = 13;
+// The GPIO pin that is oconnected to the button on the Sonoff Basic.
+const int button_gpio = 0;
+
+void relay_write(bool on) {
+    gpio_write(relay_gpio, on ? 1 : 0);
+}
+
+void led_write(bool on) {
+    gpio_write(led_gpio, on ? 0 : 1);
+}
+
+void switch_on_callback(homekit_characteristic_t *_ch, homekit_value_t on, void *context);
+
+void button_intr_callback(uint8_t gpio);
+
+homekit_characteristic_t switch_on = HOMEKIT_CHARACTERISTIC_(
+    ON, false, .callback=HOMEKIT_CHARACTERISTIC_CALLBACK(switch_on_callback)
+);
+
+void gpio_init() {
+    gpio_enable(led_gpio, GPIO_OUTPUT);
+    led_write(false);
+    gpio_enable(relay_gpio, GPIO_OUTPUT);
+    relay_write(switch_on.value.bool_value);
+    
+    gpio_set_pullup(button_gpio, true, true);
+    gpio_set_interrupt(button_gpio, GPIO_INTTYPE_EDGE_ANY, button_intr_callback);
+}
+
+void switch_on_callback(homekit_characteristic_t *_ch, homekit_value_t on, void *context) {
+    relay_write(switch_on.value.bool_value);
+}
+
+void button_intr_callback(uint8_t gpio) {
+    if (gpio_read(button_gpio) == 1) {
+        //Toggle the sonoff
+        switch_on.value.bool_value = !switch_on.value.bool_value;
+        relay_write(switch_on.value.bool_value);
+        homekit_characteristic_notify(&switch_on, switch_on.value);
+    }
+}
+
+void led_identify_task(void *_args) {
+    // We identify the Sonoff by Flashing it's LED.
+    for (int i=0; i<3; i++) {
+        for (int j=0; j<2; j++) {
+            led_write(true);
+            vTaskDelay(100 / portTICK_PERIOD_MS);
+            led_write(false);
+            vTaskDelay(100 / portTICK_PERIOD_MS);
+        }
+
+        vTaskDelay(250 / portTICK_PERIOD_MS);
+    }
+
+    led_write(false);
+
+    vTaskDelete(NULL);
+}
+
+void led_identify(homekit_value_t _value) {
+    printf("LED identify\n");
```

Comment:
Here is a thing: identify procedure is used to make sure that you have paired an accessory you thought was pairing with. Thus it's the process of identifying *accessory*. It's like it's way of saying "Here I am". In this case it's a switch. How accessory identifies itself is up to accessory: it can make sounds, blink, vibrate, etc.

The naming convention that is used across all examples is "<accessory>_<something>" (e.g. fireplace_identify or light_hue_get). In case of this accessory it should be "switch_identify", "switch_identify_task", etc. Having low level functions like "led_write" and "relay_write" is perfectly fine, but other functions should use naming convention described above. Having same naming convention simplifies understanding code for newcomers.